control/discovery.py

```diff
@@ -0,0 +1,1127 @@
+            pdu_type = pdu[0]
+            psh_flag = pdu[1]
+            ph_len = pdu[2]
+            ph_off = pdu[3]
+            package_len = pdu[4]
+            if len(self_conn.recv_buffer) < package_len:
+                lock.release()
+                return
+
+            data = self_conn.recv_buffer[:package_len]
+            self_conn.recv_buffer = self_conn.recv_buffer[package_len:]
+
+            self_conn.sq_head_ptr += 1
+            if self_conn.sq_head_ptr > SQ_HEAD_MAX:
+                self_conn.sq_head_ptr = 1
+
+            if NVME_TCP_PDU(pdu_type) == NVME_TCP_PDU.ICREQ:
+                err = self.reply_initialize(conn)
+
+            elif NVME_TCP_PDU(pdu_type) == NVME_TCP_PDU.CMD:
+                CMD1 = struct.unpack_from('<BBH', data, 8)
+                opcode = CMD1[0]
+                reserved = CMD1[1]
+                cmd_id = CMD1[2]
+
+                handle_opcode = {
+                    NVME_TCP_OPC.FABRIC_TYPE: self.reply_fabric_request,
+                    NVME_TCP_OPC.GET_LOG_PAGE: self.reply_get_log_page,
+                    NVME_TCP_OPC.IDENTIFY: self.reply_identify,
+                    NVME_TCP_OPC.SET_FEATURES: self.reply_set_feature,
+                    NVME_TCP_OPC.GET_FEATURES: self.reply_get_feature,
+                    NVME_TCP_OPC.KEEP_ALIVE: self.reply_keep_alive,
+                    NVME_TCP_OPC.ASYNC_EVE_REQ: self.store_async
+                }
+                class UnknownNVMEOpcode(BaseException):
+                    def __init__(self, opcode):
+                        super().__init__(f"unsupported opcode: {opcode}")
+                try:
+                    err = handle_opcode[opcode](conn, data, cmd_id)
+                except KeyError as e:
+                    raise UnknownNVMEOpcode(opcode) from e
+            else:
+                self.logger.error("unsupported pdu type: {pdu_type}")
+
+            if err == -1 or self_conn.shutdown_now is True:
+                del self.conn_vals[conn.fileno()]
+                self.selector.unregister(conn)
+                self.logger.debug(f"discover request from {conn} finished.")
+                conn.close()
+                lock.release()
+                return
+        lock.release()
+
+    def nvmeof_accept(self, sock, mask):
+        """Accept connection."""
+
+        conn, addr = sock.accept()
+        self.logger.debug(f"accept connection {conn} from {addr}")
+        conn.setblocking(False)
+        lock.acquire()
+        if conn.fileno() not in self.conn_vals:
+            self.conn_vals[conn.fileno()] = Connection()
+            self.conn_vals[conn.fileno()].connection = conn
+            global GLOBAL_CNLID
+            global GLOBAL_GEN_CNT
+            self.conn_vals[conn.fileno()].cnlid = GLOBAL_CNLID
+            self.conn_vals[conn.fileno()].gen_cnt = GLOBAL_GEN_CNT
+            GLOBAL_CNLID += 1
+            GLOBAL_GEN_CNT += 1
+        lock.release()
```

Comment:
* `global` is something rarely used in Python and easily replaceable by a better alternative. In this very case, a unique ID for connections, I'd rather use the [`uuid` package](https://docs.python.org/3/library/uuid.html).
* Let's use human readable var names (that's a Python essential), so `connection_id` instead of `cnlid`.
* `ctypes.Structure` allows you to initialize the struct fields.
* For connection id we can perhaps use uuids?
```suggestion
        with self.lock:
            if conn.fileno() not in self.conn_vals:
                self.conn_vals[conn.fileno()] = Connection(
                    connection=conn,
                    connection_id=uuid.uuid4(),
                    gen_cnt=self.connection_counter
                )
                self.connection_counter += 1
```